simson/common/data_transformations.py

```diff
@@ -65,6 +65,51 @@ def extrapolate_stock(
     return StockArray(**dict(stocks))
 
 
+def scale_parameter_to_future(historic_parameter : NamedDimArray, scaler : NamedDimArray) -> NamedDimArray:
+    """Scale parameter to future."""
+    dims = historic_parameter.dims.union_with(scaler.dims)
+
+    scaling_year = historic_parameter.dims.dim_list[0].items[-1]
+    n_historic = historic_parameter.shape[0]
+
+    transform_t_thist = get_subset_transformer(dims=dims, dim_letters=('t', 'h'))
+    future_parameter = get_new_array(dims, dim_letters=('t',) + historic_parameter.dims.letters[1:])
+    historic_scaler = get_new_array(dims=dims, dim_letters=('h',)+scaler.dims.letters[1:])
+    historic_share = get_new_array(dims=dims, dim_letters=historic_parameter.dims.letters)
+
+    historic_scaler[...] = scaler * transform_t_thist
+    historic_scaler[...] =  historic_scaler.maximum(1e-10)  # have a minimum value to avoid division by zero
+
+    historic_share[...] = historic_parameter / historic_scaler
+
+    scale_factor = (0.3* historic_share[...][scaling_year-1] +
+                    0.25 * historic_share[...][scaling_year-2] +
+                    0.2 * historic_share[...][scaling_year-3] +
+                    0.15 * historic_share[...][scaling_year-4] +
+                    0.1 * historic_share[...][scaling_year-5])
+
+    future_parameter[...] = scaler * scale_factor
+    future_parameter.values[:n_historic] = historic_parameter.values
+
+    return Parameter(dims=future_parameter.dims, values=future_parameter.values)
+
+
+def split_parameter_to_sectors(to_split : NamedDimArray, split_by : NamedDimArray) -> NamedDimArray:
+    """
+    Split to_split parameter data into the same sector shares as the split_by data, mainly used for in-use goods.
```

Comment:
I wonder if it makes sense to have this in sodym? It's also one of those functions where I think it would be nice to have a test, also as documentation for the different possible usecases.

Reply:
Jakob implemented this now in Sodym via the named dim array 'get shares over' function in the nda_shares branch, so I think that's nice. Maybe we can annotate it with a documentation and add a test like not having it negative and making sure the shares add up to 1 or something? Might make sense to have a discussion in general on how we want to incorporate tests, I don't have that much experience on that but happy to look it up.

Reply:
Sounds great! Yes, at the moment, tests are a nice-to-have for us. I was mostly using them to help me understand the functionalities of ndarrays in sodym. You could take a looking at the tests there, and have a go at writing a test for the get_shares_over function, if you like? Of course feel free to ask if you need any help :) But, since we don't have tests for everything anyway, then it's not something I would insist on. In general, I would like to improve the testing and increase test coverage, but I will definitely need quite a lot of time for this, and would be very happy to have some help from you if it's something you would like to learn more about :)

Reply:
Sounds good, I'll take a look! Actually I have a question already :) :  there seems to me that there are three types of tests we are using - some using the Python 'assert', some using the pydantic '@model_validator' and some direct if/else statements throwing e.g. ValueErrors. What kind of etiquette is there to when use which/which one would I use in the case above.

Reply:
Yes, true, we have different types or error catching within the code, but it depends on the situation as to which is the best to use, so not a straight forward answer, sorry. Pydantic handles a lot of error catching for us, by checking the types of the parameters we pass on initialisation of a class instance. This is nice, because it's a lot clearer and more concise that us having to write out the checks ourselves. Sometimes we have the @model_validator, if we need additional checks on initialisation that are not covered by simply specifying the type. In any case, these only perform the check on class instance initialisation, so in other cases (and for non-pydantic classes) we need to use `assert` or `if/else raises` (assert is basically just shorthand for `if/else raises`).

Okay, but the test cases are something a bit different. These are ones that we run with pytest (you can find some examples in sodym/tests, currently we don't have any for simson), which are to test that the code we have written (including the error catching) is behaving as we expect it to. Again, there are different ways to do this; we also have some `assert` statements, and `with raises`, and this depends on what we want to test, so again no straight-forward answer!

Reply:
Thank you so much for this explanation that was really helpful!

Reply:
Ok, so I looked at your code now and read up a bit on pytest. Added a new pytest for the get shares over function that replaced the above code (which jakob did in the NDArray in Sodym). Feel free to change it, it's my first try, but GitHub copilot worked quite well there so I think it should be fine.